odl/space/npy_tensors.py

```diff
@@ -0,0 +1,1992 @@
+        return long(self.data)
+
+    def __float__(self):
+        """Return ``float(self)``."""
+        return float(self.data)
+
+    def __complex__(self):
+        """Return ``complex(self)``."""
+        if self.size != 1:
+            raise TypeError('only size-1 tensors can be converted to '
+                            'Python scalars')
+        return complex(self.data[(0,) * self.ndim])
+
+    @property
+    def ufuncs(self):
+        """`NumpyTensorSpaceUfuncs`, access to Numpy style ufuncs.
+
+        Notes
+        -----
+        These ufuncs are optimized for use with `NumpyTensor`'s
+        and incur practically no overhead.
+
+        Examples
+        --------
+        >>> space = odl.rn(3)
+        >>> x = space.element([1, -2, 3])
+        >>> x.ufuncs.absolute()
+        rn(3).element([1.0, 2.0, 3.0])
+
+        Broadcasting and array-like input are supported:
+
+        >>> x.ufuncs.add(3)
+        rn(3).element([4.0, 1.0, 6.0])
+        >>> x.ufuncs.subtract([0, 0, 1])
+        rn(3).element([1.0, -2.0, 2.0])
+
+        There is also support for various reductions (sum, prod, min, max):
+
+        >>> x.ufuncs.sum()
+        2.0
+
+        The ``out`` parameter can be used to store the result:
+
+        >>> y = space.one()
+        >>> out = space.element()
+        >>> result = x.ufuncs.add(y, out=out)
+        >>> result
+        rn(3).element([2.0, -1.0, 4.0])
+        >>> result is out
+        True
+        """
+        return NumpyTensorSpaceUfuncs(self)
+
+
+def _blas_is_applicable(*args):
+    """Whether BLAS routines can be applied or not.
+
+    BLAS routines are available for single and double precision
+    float or complex data only. If the arrays are non-contiguous,
+    BLAS methods are usually slower, and array-writing routines do
+    not work at all. Hence, only contiguous arrays are allowed.
+
+    Parameters
+    ----------
+    x1,...,xN : `NumpyTensor`
+        The tensors to be tested for BLAS conformity.
+    """
+    if any(x.dtype != args[0].dtype for x in args[1:]):
+        return False
+    elif any(x.dtype not in _BLAS_DTYPES for x in args):
+        return False
+    elif not (all(x.flags.f_contiguous for x in args) or
+              all(x.flags.c_contiguous for x in args)):
+        return False
+    else:
+        return True
+
+
+def _lincomb(a, x1, b, x2, out, dtype):
+    """Raw linear combination depending on data type."""
+    # Shortcut for small problems
+    # if x1.size < 100:  # small array optimization
+    #     out.data[:] = a * x1.data + b * x2.data
+    #     return
+
+    # Need flat data for BLAS, otherwise in-place does not work
+    # Raveling must happen in fixed order for non-contiguous out,
+    # otherwise 'A' is applied to arrays, which makes the outcome
+    # dependent on their respective contiguousness.
+    if out.data.flags.f_contiguous:
+        ravel_order = 'F'
+    else:
+        ravel_order = 'C'
+
+    x1_arr = x1.data.ravel(order=ravel_order)
```

Comment:
Hm, yes. I changed it to not ravel if stuff is not contigouous in the same way. The fallbacks will handle that now. Not sure if there are proper tests for non-contiguous stuff, need to check.